Allow separate mantidqt and workbench builds

StephenSmith25 opened the pull request:
**Description of work.**
These changes allow a mantid-qt and mantid-workbench conda build package to be produced. See here for packages build on all three OSes using this branch

mantid-framework https://anaconda.org/StephenSmith25/mantid-framework
mantid-qt https://anaconda.org/StephenSmith25/mantid-qt
mantid-workbench https://anaconda.org/StephenSmith25/mantid-workbench

and see my current recipes here:

https://github.com/StephenSmith25/staged-recipes/tree/mantid/recipes

You could clone my recipe and do:
```
git clone https://github.com/StephenSmith25/staged-recipes
git checkout mantid
cd staged-recipes
cd recipes
conda build ./mantid-framework (if you want to build the others you'll have to add stephensmith25 to your conda channels.)
```


Some comments:

I don't feel like we need the exe's on windows and mac. In the conda build recipe we can define an entry point, which we can use to launch workbench. The only issue here is we miss out on the error reporter. But it does mean that using  workbench is as simple as:

```
conda install mantid-workbench
workbench
```
If we don't build the exes, our mantid-workbench package won't need a compiler or boost or qt, so It will be simplified considerably. 

I had to change the python package setup.py installation to: `pip install . --no-deps .. --vv` so it just installs directly into the conda site packages, which means you can do:

```
conda install mantid-framework
python
import mantid.kernel
import mantid.simpleapi
```



**I shall run a cmake formatter after this is merged.**

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**
You can try build the new targets with variants of the options
 ```
-DMANTID_FRAMEWORK_LIB=SYSTEM  # can be system or build
-DMANTID_QT_LIB=SYSTEM  # can be off system or build
-DENABLE_WORKBENCH=ON # can be off or on
```


<!-- Instructions for testing. -->

Fixes #32079. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

StephenSmith25 committed:
Standalone framework

StephenSmith25 committed:
adjust python library creation

StephenSmith25 committed:
separate mantidqt target

StephenSmith25 committed:
mistake in pip install

StephenSmith25 committed:
Cache path for site packages

StephenSmith25 committed:
If using system framework don't find these packages

StephenSmith25 committed:
Revise framework and qt cmake flags

StephenSmith25 committed:
Check for pyrcc with both mantidqt and workbench builds

StephenSmith25 committed:
amend find package calls

StephenSmith25 committed:
Stop finding eigen if we don't need it

StephenSmith25 committed:
Stop finding numpy if we don't need it

StephenSmith25 committed:
modify find mantidframework call

StephenSmith25 committed:
modify build options for framework and mantidqt

StephenSmith25 committed:
only find qscintilla in a mantidqt build

StephenSmith25 committed:
Skip workbench install steps if conda build

StephenSmith25 committed:
Add error message if no build components are selected

Commit:
No workbench bundle in conda build on osx

Commit:
add api link to mplcpp

Commit:
fix mistake in commenting out

Commit:
Setting correct location for resources.py

StephenSmith25 committed:
cache site packages location for linux

StephenSmith25 committed:
fix issue in qt targets due to rebase

DavidFair commented:
@StephenSmith25 Is this PR still active?
If so, could you resolve the merge conflicts and I'll look to get this in rather than it lingering around

StephenSmith25 commented:
Yep still active

DavidFair commented on CMakeLists.txt:
```diff
@@ -29,12 +29,33 @@ set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/buildconfig/CMake")
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin)
 
-option(ENABLE_WORKBENCH "Enable Qt5-based gui & components" ON)
+option(ENABLE_MANTIDPLOT "Enable Qt4-based gui & components" OFF)
```
We should nuke references to MantidPlot/Qt4 elements at this point as we see them

DavidFair commented on CMakeLists.txt:
```diff
@@ -103,7 +124,7 @@ if(COVERAGE)
 endif()
 
-if(ENABLE_WORKBENCH)
+if(ENABLE_WORKBENCH OR MANTID_QT_LIB STREQUAL "BUILD")
```
Could we move anything related to a QtBuild into it's own CMake file?

That way our top level CMakelists.txt reads something like:
```
if(ENABLE_WORKBENCH OR MANTID_QT_LIB STREQUAL "BUILD")
   add_file(buildconfig/CMake/MantidQtTargets.cmake)
```

But if it's too much work we can leave it for the future

DavidFair commented on buildconfig/CMake/Bootstrap.cmake:
```diff
@@ -179,10 +179,17 @@ endfunction()
 # Find python interpreter
 set(MINIMUM_PYTHON_VERSION 3.6)
+# If we are not building the mantid framework we don't need the numpy developer env
+if (MANTID_FRAMEWORK_LIB STREQUAL "BUILD")
```
Could we do this check once and use an internal var instead:

```
if(MANTID_FRAMEWORK_LIB STREQUAL "BUILD")
   set(MANTID_FRAMEWORK_LIB_BUILD ON)
endif()


// Some other file
if (${MANTID_FRAMEWORK_LIB_BUILD})
   // Do something
endif
```

DavidFair commented on buildconfig/CMake/CommonSetup.cmake:
```diff
@@ -80,56 +80,65 @@ add_definitions(-DBOOST_DATE_TIME_POSIX_TIME_STD_CONFIG)
+
+endif()
+# if we are building the framework we will need these libraries.
+if (MANTID_FRAMEWORK_LIB STREQUAL "BUILD")
```
Ditto about splitting this into it's own Framework_Lib cmake file (if it's trivial)

DavidFair reviewed (changes requested):
Some high level readability improvements

We also need to add `conda-build` to the mantid-developer conda env too
Edit and `conda verify` too

I'm still building away on OSX with the various recipes...

StephenSmith25 committed:
Fix some rebase issues

StephenSmith25 committed:
Use internal build variables rather than strcompares

Adds internal build variables, rather than using strcmp to check if we are building certain components

DavidFair reviewed (approved):
LGTM, we can tidy up the other minor bits as we go, but I'd like to get this big change in

I've re-tested with:
mantid-framework: Linux, OSX, Windows
mantid-qt: Windows
mantid-workbench: Windows
